server/registry/internal/storage/delete.go

```diff
@@ -24,30 +25,57 @@ import (
 	"gorm.io/gorm"
 )
 
+func sum(counts []int64) int64 {
+	var s int64
+	for _, v := range counts {
+		s += v
+	}
+	return s
+}
+
+func reason(counts []int64, tables []interface{}) string {
+	r := ""
+	for i, v := range counts {
+		if v > 0 {
+			r += fmt.Sprintf(" %T:%d", tables[i], v)
+		}
+	}
+	return r
+}
+
 func (c *Client) DeleteProject(ctx context.Context, name names.Project, cascade bool) error {
 	err := c.db.WithContext(ctx).Transaction(func(tx *gorm.DB) error {
-		var count int64
-		for _, model := range []interface{}{
+		tables := []interface{}{
 			models.Project{},
```

Comment:
Did we consider moving the project deletion into the next loop (after the count check) instead of counting each deletion separately? That approach still allows us to point out a specific child type and how many rows are blocking the deletion.

The question is about how much value we get from having multiple child types/counts vs. a single child type/count, and whether it's worth the added code complexity. I definitely agree we should point out at least one child type/count.

We can add the number of rows affected per query to our database logs. Just add something like `"count": rowsAffected,` after line 58 below. Then in cases where we really want to know the full picture we can check the logs to see which tables have rows blocking the delete.

https://github.com/apigee/registry/blob/2c94ac7bf013279f5c9747c2b6bc37587181a189/server/registry/internal/storage/logger.go#L54-L68

Reply:
Thanks, I was aiming to provide detail that would be useful to the caller - a next step further in that direction would be to list the specific resources that are blocking, but per-resource counts seemed to be decent middle ground. I'm not sure that logs or making subsequent API calls to find child resources would be as effective at providing easy-to-use diagnostic messages, particularly in restricted environments like CI runners. But we could certainly revisit this. I would be happy to track an issue about this but for now suggest we tag #636, which is likely to rearrange/replace this code.

Looking at the root cause of the problems we saw in this PR, we might want to find ways to invalidate page tokens when deletions are made to collections. Perhaps storing something like time-of-last-deletion for each collection would allow us to recognize when a deletion has occurred in the middle of pagination and return a clear and actionable error message ("don't delete resources when you are iterating"). That might allow one user's deletion to break the iterations of another, but that is probably better than failing silently.